stdlib/filename.mli

```diff
@@ -165,3 +165,33 @@ val quote : string -> string
     with programs that follow the standard Windows quoting
     conventions.
  *)
+
+val quote_command :
+       string -> ?stdin:string -> ?stdout:string -> ?stderr:string
+              -> string list -> string
+(** [quote_command cmd args] returns a quoted command line, suitable
+    for use as an argument to {!Sys.command}, {!Unix.system}, and the
+    {!Unix.open_process} functions.
+
+    The string [cmd] is the command to call.  The list [args] is
+    the list of arguments to pass to this command.  It can be empty.
+
+    The optional arguments [?stdin] and [?stdout] and [?stderr] are
+    file names used to redirect the standard input, the standard
+    output, or the standard error of the command.
+    If [~stdin:f] is given, a redirection [< f] is performed and the
+    standard input of the command reads from file [f].
+    If [~stdout:f] is given, a redirection [> f] is performed and the
+    standard output of the command is written to file [f].
+    If [~stderr:f] is given, a redirection [2> f] is performed and the
+    standard error of the command is written to file [f].
+    If both [~stdout:f] and [~stderr:f] are given, with the exact
+    same file name [f], a [2>&1] redirection is performed so that the
+    standard output and the standard error of the command are interleaved
+    and redirected to the same file [f].
+
+    Under Unix and Cygwin, the command, the arguments, and the redirections
+    if any are quoted using {!Filename.quote}, then concatenated.
+    Under Win32, additional quoting is performed as required by the
+    [cmd.exe] shell that is called by {!Sys.command}.
+*)
```

Comment:
Could you make it explicit that quote_command can raise (and in which condition)?

Also, I'm not fully convinced that `Invalid_arg` should be raised, and not `Failure`.  `Invalid_arg` reports programming errors, and should not usually be caught; in this case, this would mean that the caller needs to validate the redirection filenames, since they could easily be derived from some program inputs.  The validation is not overly difficult, but not straightforward and it is platform dependent.  In practice, programmers will not implement the validation.

If one decides to raise `Failure`, one could go away with a more informal contract with the client ("The function will raise Failure if some arguments cannot be properly escaped on the current platform"), and programmers might actually decide to catch the exception.

@dbuenzli : as our champion for Failure vs Invalid_arg debate, feel free to comment.

Reply:
It should be `Failure` rather than `Invalid_argument` because it's a failure of the library itself rather than the program: in principle, every file name should be quotable, it's just that we don't (yet) know how to do it on Windows.